### PR TITLE
Fix podWorkloadBuilder (#48854)

### PR DESCRIPTION
**Pull request by zengyuxing007**
**Please provide a description of this PR:**

Fixes https://github.com/istio/istio/issues/48854

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:202]: I don't get why we need pending. If its 'pending' its not running so it cannot send traffic?

Is it just that this makes sure we have the workload before the pod starts instead of ~100ms after?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:202]: i think we should support init container btw

Comment (bleggett) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:202]: > if it's "pending" doesn't that mean the init container has not started? 

No, pod phase of "pending" means init container haven't finished.

So agree, probably not the right metric.

> A Pod cannot be Ready until all init containers have succeeded. The ports on an init container are not aggregated under a Service. A Pod that is initializing is in the Pending state but should have a condition Initialized set to false.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:202]: TIL!

"but one or more of the containers has not been started"

So technically it may not have started yet, but probably good enough. We should make sure its assigned to a node though -- it will start in pending but no node

**Review by howardjohn — COMMENTED** [review]
LGTM if we add the node check (and Ben is ok with this)

File: pilot/pkg/serviceregistry/kube/controller/ambient/helpers.go
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/ambient/helpers.go:137]: Can we look at a more granular state than the `Phase`? To quote the docs:

> The phase of a Pod is a simple, high-level summary of where the Pod is in its lifecycle. The phase is not intended to be a comprehensive rollup of observations of container or Pod state, nor is it intended to be a comprehensive state machine.


Can we instead use the PodCondition: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions?

Specifically, `PodScheduled` and/or `PodReadyToStartContainers`

(what we really want here is `PodReadyToStartContainers` I think, as it guarantees network readiness, but that is currently a k8s beta feature - and since we have a CNI plugin strictly speaking we don't have to wait on k8s to tell us about network readiness)

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]: This should be fine.

Pending but with IP == CNI has had a shot already, so we are waiting on processes in the pod to spawn.

Pending with no IP == we can't guarantee CNI has had a shot already.

What we are worried about in ambient is making sure the pod netns has actually been created before we try to jump into it.

"Pod got an IP" is a good enough proxy for that, I think.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]: I think we basically want: pod scheduled to node, assigned IP, and its not terminated.

Which we could do with: `ip != "" && nodeName != "" && phase in [pending, running]` or equivalent through conditions

Comment (bleggett) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]: it's not the end of the world if it's terminated, but checking would avoid some extra churn/excess failure messages.

Comment (zengyuxing007) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]:
```yaml
...
  status
    phase: Pending  Or Running
    podIP: 10.0.174.216
    podIPs:
    - ip: 10.0.174.216
 ```
 
 If  pod status is Pending (not terminated or failed or completed or unknow) but have been assigned IP  which mean that  pod have be scheduled to a node and is initialing.  
 
 There were related filters before, and I added the podPending judgment, which I think is consistent with what we're describing above.  This is the complete code for this logic after the merge，

```go
func (a *index) podWorkloadBuilder(
	MeshConfig krt.Singleton[MeshConfig],
	AuthorizationPolicies krt.Collection[model.WorkloadAuthorization],
	PeerAuths krt.Collection[*securityclient.PeerAuthentication],
	Waypoints krt.Collection[Waypoint],
	WorkloadServices krt.Collection[model.ServiceInfo],
) func(ctx krt.HandlerContext, p *v1.Pod) *model.WorkloadInfo {
	return func(ctx krt.HandlerContext, p *v1.Pod) *model.WorkloadInfo {
		if (!IsPodRunning(p) && !IsPodPending(p)) || p.Spec.HostNetwork {
			return nil
		}
		podIP, err := netip.ParseAddr(p.Status.PodIP)
		if err != nil {
			// Is this possible? Probably not in typical case, but anyone could put garbage there.
			return nil
		}
```

Comment (zengyuxing007) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]: > it's not the end of the world if it's terminated, but checking would avoid some extra churn/excess failure messages - yeah `PodReadyToStartContainers` is ideal, in contexts where it's available.

Istio now support kubernetes 1.26, 1.27, 1.28, 1.29 (https://istio.io/latest/docs/releases/supported-releases/), 

but  `PodReadyToStartContainers` was only [made default in kubernetes 1.29](https://kubernetes.io/blog/2023/12/19/pod-ready-to-start-containers-condition-now-in-beta/), which can lead to unfriendly support for other lower kubernetes versions.

Comment (bleggett) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]: If we are going to use a custom heuristic here (Pending + some_other_conditions) then it would be good to add a comment explaining that in detail to this spot so we don't forget what and why.

We should be able to replace all those checks with a single check against `PodReadyToStartContainers` eventually, but for now I'm fine with !running + is_pending + has_ip

Comment (zengyuxing007) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads_test.go:77]: Ok, thanks for your reply.

I've added the relevant notes, please check it again~

**Review by bleggett — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]
LGTM

**Comment by bleggett** [comment]
@zengyuxing007 did we want to cherrypick this into 1.21?

**Comment by bleggett** [comment]
Actually it's built on top of the `krt` stuff which is not 1.21 - so maybe not.

**Comment by zengyuxing007** [comment]
> Actually it's built on top of the `krt` stuff which is not 1.21 - so maybe not.

yes ~

**Pull request merged**
